File: frontend/src/utils/geo.js
import { SDM_COLLEGE } from "../services/routeConfig";

export const campusCenter = [SDM_COLLEGE.lat, SDM_COLLEGE.lng];

export function haversineMeters(a, b) {
  const radius = 6371000;
  const toRad = (value) => (value * Math.PI) / 180;
  const dLat = toRad(b.lat - a.lat);
  const dLng = toRad(b.lng - a.lng);
  const lat1 = toRad(a.lat);
  const lat2 = toRad(b.lat);
  const value =
    Math.sin(dLat / 2) ** 2 + Math.cos(lat1) * Math.cos(lat2) * Math.sin(dLng / 2) ** 2;
  return 2 * radius * Math.asin(Math.sqrt(value));
}

export function nextPoint(polyline, tick) {
  if (!polyline?.length) return { lat: campusCenter[0], lng: campusCenter[1] };
  const index = tick % polyline.length;
  const [lat, lng] = polyline[index];
  return { lat, lng };
}

File: backend/app/services/geo.py
from math import asin, cos, radians, sin, sqrt


EARTH_RADIUS_M = 6371000


def haversine_m(a: dict, b: dict) -> float:
    lat1, lng1 = radians(a["lat"]), radians(a["lng"])
    lat2, lng2 = radians(b["lat"]), radians(b["lng"])
    dlat = lat2 - lat1
    dlng = lng2 - lng1
    value = sin(dlat / 2) ** 2 + cos(lat1) * cos(lat2) * sin(dlng / 2) ** 2
    return 2 * EARTH_RADIUS_M * asin(sqrt(value))


def calculate_speed_kmph(previous: dict | None, current: dict, seconds: float) -> float:
    if not previous or seconds <= 0:
        return 0
    meters = haversine_m(previous, current)
    return round((meters / seconds) * 3.6, 1)


def nearest_point_distance_m(point: dict, points: list[dict]) -> float:
    if not points:
        return float("inf")
    return min(haversine_m(point, candidate) for candidate in points)
